### PR TITLE
Migrate to tf.module  and add support for SavedModels

### PR DESCRIPTION
**Pull request by sdmonov**
This PR changes the way the converter executes the computational graph. It moves to using tf.function / tf.module and AutoGraph.

### CODE REVIEW

**Comment by sdmonov** [comment]
cc @chinhuang007

**Comment by chinhuang007** [comment]
Generally looking good. Please 1. fix the LGTM alerts like unused local variables 2. comment on some failed test cases in running test_onnx_backend.py such as test_nonmaxsuppression_two_classes_cpu and test_nonmaxsuppression_two_batches_cpu. If these are isolated in the operator, please open an issue to track them. Thanks.

**Comment by sdmonov** [comment]
Tests which still have issues:
test_onnx_backend.py:

test_gru_defaults_cpu
test_gru_seq_length_cpu
test_gru_with_initial_bias_cpu
test_lstm_defaults_cpu
test_lstm_with_initial_bias_cpu
test_lstm_with_peepholes_cpu
test_rnn_seq_length_cpu
test_simple_rnn_defaults_cpu
test_simple_rnn_with_initial_bias_cpu
test_operator_symbolic_override_nested_cpu
test_nonmaxsuppression_two_batches_cpu
test_resize_downsample_nearest_cpu
test_resize_nearest_cpu
test_resize_upsample_linear_cpu
Many of these issues are related to each other (fixing few should fix all)

test_node.py:
test_erf - the result is very close to the test - need to re evalute the test_similar precision?

test_dynamic_shape.py:
test_compress - looks simple some type issues

**Comment by chinhuang007** [comment]
@sdmonov I merged the current master with your PR and found a few errors in test_node.py. 
The first is coming from onnx_tf/handlers/backend/pad.py. The function process_pos_pads works in python but not with tf.function, complaining variable constant_values is undefined. So a simple solution is to add a new parameter to process_pos_pads and process_neg_pads and specifically pass it in. Can you please add the change to this PR? Since the current code works, I don't feel necessary to open a separate PR.
The second is issue is in test_if. @winnietsang will also provide a fix to be included in this PR.

**Comment by chinhuang007** [comment]
@sdmonov I fixed another problem in test_model.py and decided to push a commit to include both of my fixes. So you can ignore my previous comment.

**Comment by chinhuang007** [comment]
Made another commit to fix problems in test_node.py and test_dynamic_shape.py for topk, argmax, argmin, and compress tests. So our local tests are clean now for opset 11.

**Comment by chinhuang007** [comment]
Saw additional errors from test_onnx_backend.py with the latest onnx.
- test_dynamicquantizelinear_expanded_cpu
- test_dynamicquantizelinear_max_adjusted_expanded_cpu
- test_dynamicquantizelinear_min_adjusted_expanded_cpu

Please take a look. Thanks.

File: onnx_tf/backend_rep.py [hunk 1]
```diff
@@ -95,15 +91,4 @@ def export_graph(self, path):
...
-    for output_name in self.outputs:
-      meaningful_names[self.tensor_dict[output_name].name.replace(':0', '')] = output_name
-    for node in graph_proto.node:
-      if node.name in meaningful_names.keys():
-        node.name = meaningful_names[node.name]
-
-    file = open(path, "wb")
-    file.write(graph_proto.SerializeToString())
-    file.close()
+    tf.saved_model.save(self.tf_module, path, signatures=self.tf_concrete_function)
```
Comment (chinhuang007) [inline onnx_tf/backend_rep.py:94]: tf_concrete_function doesn't seem to have been assigned anything. The saved model won't work without input signatures.

File: onnx_tf/backend_tf_module.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      input_dict_items = self.backend._onnx_initializer_to_input_dict_items(
+          self.graph_def.initializer)
+      initialized = {init.name for init in self.graph_def.initializer}
+    else:
+      input_dict_items = []
+      initialized = set()
+
+    tensor_dict.update(input_dict_items)
+
+    output = []
```
Comment (chinhuang007) [inline onnx_tf/backend_tf_module.py:28]: This line is not needed.

File: onnx_tf/backend_tf_module.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    for node in self.graph_def.node:
+      onnx_node = OnnxNode(node)
+      output_ops = self.backend._onnx_node_to_tensorflow_op(
+          onnx_node, tensor_dict, self.handlers, opset=self.opset, strict=self.strict)
+      curr_node_output_map = dict(zip(onnx_node.outputs, output_ops))
+      tensor_dict.update(curr_node_output_map)
+
+    outputs = [tensor_dict[output] for output in self.outputs]
```
Comment (chinhuang007) [inline onnx_tf/backend_tf_module.py:36]: Shouldn't there be a class variable for outputs? The use of self.outputs seems strange. Do you assume outputs variable is added by TF with tf.function?

**Comment by chinhuang007** [comment]
@sdmonov Looks good and works as expected. A couple of minor changes if possible:
1. fix the warnings reported by the LGTM tool, https://github.com/onnx/onnx-tensorflow/pull/603/checks?check_run_id=1088939652
2. add some filtering with comments to test/backend/test_onnx_backend.py so the 4 lstm, gru, and rnn tests can be filtered out for CI to stay green

**Review by chinhuang007 — APPROVED** [review]
LGTM!

File: onnx_tf/handlers/backend/non_max_suppression.py [hunk 4]
```diff
@@ -40,7 +42,6 @@ def _common(cls, node, **kwargs):
       boxes_t = tf.concat([y1, x1, y2, x2], 1)
       boxes = tf.transpose(boxes_t, perm=[0, 2, 1])
 
-    @tf.function
```
Comment (khatchad) [inline onnx_tf/handlers/backend/non_max_suppression.py:53]: Wonder why `@tf.function` was removed here.

**Pull request merged**